// angular-neo4j/bundles/angular-neo4j.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular/forms'), require('neo4j-driver/lib/browser/neo4j-web')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common', '@angular/forms', 'neo4j-driver/lib/browser/neo4j-web'], factory) :
	(factory((global['angular-neo4j'] = {}),global.ng.core,global.ng.common,global.ng.forms,global.neo4j));
}(this, (function (exports,core,common,forms,neo4j) { 'use strict';

neo4j = neo4j && neo4j.hasOwnProperty('default') ? neo4j['default'] : neo4j;

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var AngularNeo4jService = /** @class */ (function () {
    function AngularNeo4jService() {
    }
    /**
     * Create a new driver connection
     * @param {?} url
     * @param {?} username
     * @param {?} password
     * @param {?=} encrypted
     * @return {?}
     */
    AngularNeo4jService.prototype.connect = function (url, username, password, encrypted) {
        var _this = this;
        if (encrypted === void 0) { encrypted = true; }
        return new Promise(function (resolve, reject) {
            try {
                var /** @type {?} */ auth = neo4j.auth.basic(username, password);
                if (username && password && encrypted) {
                    _this.driver = neo4j.driver(url, auth, { encrypted: encrypted });
                }
                else if (username && password) {
                    _this.driver = neo4j.driver(url, auth);
                }
                else {
                    _this.driver = neo4j.driver(url);
                }
                resolve(_this.driver);
            }
            catch (e) {
                reject(e);
            }
        });
    };
    /**
     * Close a driver connection
     * @return {?}
     */
    AngularNeo4jService.prototype.disconnect = function () {
        if (this.driver) {
            this.driver.close();
            this.driver = null;
        }
    };
    /**
     * Get the last instantiated driver instance
     * @return {?}
     */
    AngularNeo4jService.prototype.getDriver = function () {
        if (!this.driver) {
            throw new Error('A connection has not been made to Neo4j. You will need to run `connect(url, username, password)` before you can get the current driver instance');
        }
        return this.driver;
    };
    /**
     * Create a new driver session
     * @return {?}
     */
    AngularNeo4jService.prototype.getSession = function () {
        if (!this.driver) {
            throw new Error('A connection has not been made to Neo4j. You will need to run `connect(url, username, password)` before you can create a new session');
        }
        return this.driver.session();
    };
    /**
     * Run a query on the current driver
     * @param {?} query
     * @param {?=} params
     * @return {?}
     */
    AngularNeo4jService.prototype.run = function (query, params) {
        var _this = this;
        if (params === void 0) { params = null; }
        var /** @type {?} */ session = this.getSession();
        return session.run(query, params).then(function (results) {
            session.close();
            return results.records.map(function (record) {
                var /** @type {?} */ r = [];
                for (var /** @type {?} */ i = 0; i < record.length; i++) {
                    r[i] = _this.processRecord(record.get(i));
                }
                return r;
            });
        }, function (err) {
            session.close();
            throw err;
        });
    };
    /**
     * @param {?} integer
     * @return {?}
     */
    AngularNeo4jService.prototype.processInteger = function (integer) {
        if (integer.constructor.name === 'Integer') {
            return integer.toNumber();
        }
        return integer;
    };
    /**
     * @param {?} record
     * @return {?}
     */
    AngularNeo4jService.prototype.processRecord = function (record) {
        var _this = this;
        if (!record) {
            return null;
        }
        if (record.constructor.name === 'Integer') {
            return record.toNumber();
        }
        if (record.constructor.name === 'Path') {
            record.start.identity = this.processInteger(record.start.identity);
            record.end.identity = this.processInteger(record.end.identity);
            record.segments = record.segments.map(function (segment) {
                segment.start.identity = _this.processInteger(segment.start.identity);
                segment.end.identity = _this.processInteger(segment.end.identity);
                segment.relationship.identity = _this.processInteger(segment.relationship.identity);
                segment.relationship.start = _this.processInteger(segment.relationship.start);
                segment.relationship.end = _this.processInteger(segment.relationship.end);
                return segment;
            });
            return record;
        }
        if (record.constructor.name === 'Relationship') {
            record.identity = this.processInteger(record.identity);
            record.start = this.processInteger(record.start);
            record.end = this.processInteger(record.end);
            return record;
        }
        if (record.constructor.name === 'Node') {
            record.identity = this.processInteger(record.identity);
            return record;
        }
        return record;
    };
    return AngularNeo4jService;
}());
AngularNeo4jService.decorators = [
    { type: core.Injectable, args: [{
                providedIn: 'root'
            },] },
];
/** @nocollapse */
AngularNeo4jService.ctorParameters = function () { return []; };
/** @nocollapse */ AngularNeo4jService.ngInjectableDef = core.defineInjectable({ factory: function AngularNeo4jService_Factory() { return new AngularNeo4jService(); }, token: AngularNeo4jService, providedIn: "root" });
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var AngularNeo4jComponent = /** @class */ (function () {
    /**
     * @param {?} fb
     * @param {?} neo4j
     */
    function AngularNeo4jComponent(fb, neo4j$$1) {
        this.fb = fb;
        this.neo4j = neo4j$$1;
    }
    /**
     * @return {?}
     */
    AngularNeo4jComponent.prototype.ngOnInit = function () {
        this.createForms();
    };
    /**
     * @return {?}
     */
    AngularNeo4jComponent.prototype.createForms = function () {
        this.loginForm = this.fb.group({
            url: 'bolt://localhost:7687',
            username: '',
            password: '',
            encrypted: true
        });
        this.queryForm = this.fb.group({
            query: '',
            params: ''
        });
    };
    /**
     * @return {?}
     */
    AngularNeo4jComponent.prototype.connect = function () {
        var _this = this;
        var /** @type {?} */ formModel = this.loginForm.value;
        this.neo4j
            .connect(formModel.url, formModel.username, formModel.password, formModel.encrypted)
            .then(function (driver) {
            if (driver) {
                console.log("Successfully connected to " + _this.loginForm.value.url);
            }
        });
    };
    /**
     * @return {?}
     */
    AngularNeo4jComponent.prototype.disconnect = function () {
        this.neo4j.disconnect();
    };
    /**
     * @return {?}
     */
    AngularNeo4jComponent.prototype.run = function () {
        var _this = this;
        var /** @type {?} */ formModel = this.queryForm.value;
        var /** @type {?} */ params = {};
        if (formModel.params) {
            try {
                params = JSON.parse(formModel.params);
            }
            catch (err) {
                console.error(err);
            }
        }
        this.neo4j.run(formModel.query, params).then(function (res) {
            _this.results = res;
            console.log(res);
        });
    };
    return AngularNeo4jComponent;
}());
AngularNeo4jComponent.decorators = [
    { type: core.Component, args: [{
                selector: 'angular-neo4j',
                template: "<form [formGroup]=\"loginForm\" (ngSubmit)=\"connect()\"> <label>URL: <br /> <input formControlName=\"url\" placeholder=\"bolt://host:port\"> </label> <br /> <br /> <label>Username: <br /> <input formControlName=\"username\"> </label> <br /> <br /> <label>Password: <br /> <input formControlName=\"password\"> </label> <br /> <br /> <label> <input formControlName=\"encrypted\" type=\"checkbox\"> Encrypted </label> <br /> <br /> <button type=\"submit\">Connect</button> <button type=\"button\" (click)=\"disconnect()\">Disconnect</button> </form> <br /> <br /> <form [formGroup]=\"queryForm\" (ngSubmit)=\"run()\"> <label>Cypher Query: <br /> <textarea formControlName=\"query\" placeholder=\"MATCH (o:Object {attrib: {value}}) RETURN o\"></textarea> </label> <br /> <br /> <label>Parameters JSON: <br /> <textarea formControlName=\"params\" placeholder='{ \"value\": \"value\" }'></textarea> </label> <br /> <br /> <button type=\"submit\">Run</button> </form> <br /> <br /> Results: <pre> {{results | json}} </pre>",
                styles: [""]
            },] },
];
/** @nocollapse */
AngularNeo4jComponent.ctorParameters = function () { return [
    { type: forms.FormBuilder, },
    { type: AngularNeo4jService, },
]; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var AngularNeo4jModule = /** @class */ (function () {
    function AngularNeo4jModule() {
    }
    /**
     * @return {?}
     */
    AngularNeo4jModule.forRoot = function () {
        return {
            ngModule: AngularNeo4jModule,
            providers: [AngularNeo4jService]
        };
    };
    return AngularNeo4jModule;
}());
AngularNeo4jModule.decorators = [
    { type: core.NgModule, args: [{
                imports: [common.CommonModule, forms.ReactiveFormsModule],
                exports: [AngularNeo4jComponent],
                providers: [AngularNeo4jService],
                declarations: [AngularNeo4jComponent]
            },] },
];

exports.AngularNeo4jModule = AngularNeo4jModule;
exports.AngularNeo4jComponent = AngularNeo4jComponent;
exports.AngularNeo4jService = AngularNeo4jService;

Object.defineProperty(exports, '__esModule', { value: true });

})));
